Clarify how to make progress on a PR (#745)
open-telemetry/opentelemetry-specification: CONTRIBUTING.md

```bash
make misspell-correction
```

## Pull Request

### How to Send Pull Requests

Everyone is welcome to contribute code to `OpenTelemetry Specification` via
GitHub pull requests (PRs).

> To create a new PR, check out a new branch, make modifications and push the branch:

Reviewer: We should mention about forking this repo first, then creating branch/pushing to that fork.

```sh
git checkout -b ${USERNAME}/feature_branch_name
# edit files
git commit -m "Description"
git push ${USERNAME}/feature_branch_name
```

Open a pull request against the master branch.

### How to Receive Comments

* If the PR is not ready for review, please put `[WIP]` in the title or mark it
  as [`draft`](https://github.blog/2019-02-14-introducing-draft-pull-requests/).
* Make sure CLA is signed and CI is clear.

### How to Get PR Merged

A PR is considered to be **ready to merge** when:

> * It has received two approvals from the [code owners](./CODEOWNERS) (at

Reviewer 2: How is different this to https://github.com/open-telemetry/opentelemetry-specification/blob/master/issue-management.md#closing-an-issue?

Sounds like some of the content here can be added there instead?

Author: I think the `issue-management.md` document is targeting all the repositories instead of the spec repo, and it is quite outdated (e.g. the document mentioned a tag `will-not-fix` which doesn't exist at all - the correct tag is `wontfix`). Changing it would be a bigger scope than the purpose of this PR.

I'm happy to send a separate PR to clean up the `issue-management.md` or even consolidate the issue and PR management and apply them across the spec and SDKs.

The issue part can be updated with something like this:

* Review criteria:
  * For interface and design changes: 2 approvals - which must be from reviewers
    who work at different companies than the Collaborator.
  * Major feedbacks are resolved.
  * It has been at least two working days since the last modification on the
    issue description (except for the trivial updates, such like typo, cosmetic,
    rebase, etc.). This gives people reasonable time to review.
  * Trivial change (typo, cosmetic etc.) on the issue description doesn't have
    to wait for two days.

  different companies).
* Major feedbacks are resolved.
* It has been at least two working days since the last modification (except for
  the trivial updates, such like typo, cosmetic, rebase, etc.). This gives
  people reasonable time to review.
* Trivial change (typo, cosmetic, CI improvement, etc.) doesn't have to wait for
  one day.

Any [spec
approver](https://github.com/orgs/open-telemetry/teams/specs-approvers) can
merge the PR once it is **ready to merge**.

### If a PR is stuck

> If a PR has been stuck for more than 2 weeks, the owner should bring it to the

Reviewer 2: +1

> [OpenTelemetry Specification SIG

Reviewer 3: Is this the only way to unstuck a PR? How about mentioning gitter and people that would be good to ping? Relying on the SIG to get unstuck seems to have the problems of

- SIG meeting agenda will get bogged down by PRs
- SIG meeting is timezone-unfriendly for a lot of the world

Author: I guess it is assumed that people will try their best to resolve the conflict in PR comments / gitter chat / offline chat?
Happy to add some clarification here if it is not very obvious.

Reviewer 3: People might not know about gitter (actually I couldn't find a gitter link on the home page readme or contributing.md), or they may be shy about it. I think linking it here with encouragement to ping there can help

meeting](https://github.com/open-telemetry/community#cross-language-specification).

> If a PR has been opened for more than 6 months, it will be closed.

Reviewer: Al alternate PR is adding bots with 14 days https://github.com/open-telemetry/opentelemetry-specification/pull/744/files#diff-a1810534b70b5e9fffb625daf4f0d619R15

This text and the bot should state the same "idle period"

Author: That's a good point. We should do this in other language SIGs, too!

Author: I've removed this line.

```bash
make misspell-correction
```

> ## Pull Request

Reviewer 4: ```suggestion
## Pull Requests
```

> ### How to send PR

Reviewer 4: ```suggestion
### How to create a PR
```

> Everyone is welcome to contribute to `OpenTelemetry Specification` via GitHub

Reviewer 4: ```suggestion
Everyone is welcome to contribute to the OpenTelemetry specification via GitHub
```

pull requests (PRs).

> To create a new PR, fork the project in GitHub and clone the upstream repo:

Reviewer 4: We may want to link to https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request

Author: Good point! Added.

```sh
git clone https://github.com/open-telemetry/opentelemetry-specification.git
```

Add your fork as a remote:

```sh
git remote add fork https://github.com/YOUR_GITHUB_USERNAME/opentelemetry-specification.git
```

Check out a new branch, make modifications and push the branch to your fork:

```sh
$ git checkout -b feature
# edit files
$ git commit
$ git push fork feature
```

Open a pull request against the main `opentelemetry-specification` repo.

> ### How to receive comments

Reviewer: minor: 
The contents under this heading doesn't really address "how to receive comments"?

Author: It is necessary requirements but not aiming to be complete.

Author: E.g. we don't encourage approvers/maintainers to review/comment if CLA is not signed.

Reviewer 4: I think @cijothomas was not arguing that the contents of this section aren't useful but rather that the heading does not fit the content. How about:

```suggestion
### How to create your PR
```

and the above section should be "How to prepare your changes" instead of "How to send PR"

Reviewer 4: > E.g. we don't encourage approvers/maintainers to review/comment if CLA is not signed.

This is not at all how I read this section. If that was the intention, it should be clarified.

Author: I've merged this section with the above section so we don't run into such struggle.

Author: > > E.g. we don't encourage approvers/maintainers to review/comment if CLA is not signed.
> 
> This is not at all how I read this section. If that was the intention, it should be clarified.

I've added some clarification, please take a look.

> * If the PR is not ready for review, please put `[WIP]` in the title or mark it

Reviewer: Instead of advising both approaches, why not just say "mark as Draft"? Draft is intended to solve this exact problem, right?

Author: I think they have different semantics and I've seen people use both of them in combination.
A draft PR won't trigger CI, if I'm making CI changes I probably want to mark my PR as `[WIP]` to test things out, but gives people a single that I don't expect them to take the PR review yet.

Reviewer 4: > A draft PR won't trigger CI

It will in spec. In fact that's one way I have been using drafts: Create a draft, once CI passes (after any fixes), click "Ready for review".

Reviewer: @reyang If draft PR does trigger PRs, then I am more in favor of avoiding "WIP" text in favor of Github recommended way of using Drafts to indicate "this is WIP".

Author: I don't have strong opinion on this so I've removed `[WIP]' and only mention `draft`.

Please make sure CLA is signed and CI is clear. We don't expect people to review
and comment on a PR that doesn't have CLA signed.

> ### How to get PR merged

Reviewer 4: ```suggestion
### How to get your PR merged
```

A PR is considered to be **ready to merge** when:

> * It has received two approvals from the [code owners](./.github/CODEOWNERS) (at

Reviewer: just realized this doesn't clarify whether the 2 approver's company should be different from the PR authors company.

Reviewer 4: Also, is this a new requirement? If so, this PR goes far beyond "clarification".

Author: Strictly speaking it is new as it was never described/clarified in the doc.
However there has been such practice in many other places, for example [here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/issue-management.md#closing-an-issue).

Author: I don't think we should clarify this as we're going to get more trouble:
* What if someone doesn't work for any company (e.g. freelancer).
* What is a definition of a company, do we count subsidiary companies as the same one or not.
* What if someone approved the PR while working for company A, and then moved to company B before that PR gets merged.

There are infinite of possibility and I can always give corner cases to challenge the wording. I think we should step back and not try to be a lawyer.

Reviewer 4: @reyang I think the cases you mentioned can be considered edge cases. However, it is usually the case that the reviewer and the creator of the PR are not the same person. 😃 Thus I think we should define that case (no strong opinion though).

Author: I've added some clarification, please see if it makes things better 😄

Reviewer 4: > if approvals are from only one company, it won't count

So this means, as soon as there are at least two approvals that are not all from the same company, the PR is approved "enough", regardless of the company affiliation of the PR author?

  different companies).

> * Major feedbacks are resolved.

Reviewer: "Major" can be interpreted differently - Should we instead say "no active "request-changes" from any codeowners? This forces code owners to make sure they put "request-changes" when they make feedbacks which they consider as blocking. If the feedback was more like "nice to haves", then they won't "request-changes", and won't block merges due to different interpretation of "major"

Reviewer 4: There should also be guidelines for reviewers when to give "request changes". Or we could introduce a recommendation to mark major comments a certain way, e.g. by putting an checkbox next to it.

- [ ] (Example:) Please resolve this before merging.

Author: I've changed this to "There is no `request changes` from the [code owners](./.github/CODEOWNERS)".

* It has been at least two working days since the last modification (except for
  the trivial updates, such like typo, cosmetic, rebase, etc.). This gives
  people reasonable time to review.

> * Trivial change (typo, cosmetic, CI improvement, etc.) doesn't have to wait for

Reviewer 4: ```suggestion
* Trivial changes (typos, cosmetic changes, CI improvements, etc.) don't have to wait for
```

  two days.

Any [spec
approver](https://github.com/orgs/open-telemetry/teams/specs-approvers) can

> merge the PR once it is **ready to merge**.

Reviewer 4: Is this going to be changed? Right now only the TC can merge AFAIK. Or can all the specs-approvers (but not specs-trace-approvers) merge?

Author: This is something I want to get feedback.

Reviewer 4: I have the feeling that some PRs are held back by maintainers not being interested in them despite the PRs having enough interested approvers. Also I'm a trace-approver but no maintainer myself 😃  So personally I'm in favor of this change, but if we had some statistics on PRs (what percentage of their lifetime do they spend waiting for a maintainer to merge them after having enough approvals?) that would be even better.

Reviewer 2: @reyang Please remove this part to reduce scope. We can follow up later. Once this is done, we are ready to merge.

Reviewer 5: Yes, I think this kind of changes needs a charter update, and should be discussed in the community repo.

Reviewer 2: Ping @reyang

Reviewer 2: Ping @reyang ;)

Author: Updated.

If a PR has been stuck (e.g. there are lots of debates and people couldn't agree

> on each other), the owner should try to get people aligned by:

Reviewer 3: Thanks!

> * Consolidating the perspectives and putting a summary in the PR.

Reviewer 4: In the PR description? In a comment? In a comment linked in the PR description? From experience I'd suggest the latter since many people skip the PR description, but in a long conversation it's hard to find a particular comment too.

Author: I guess we can leave this flexibility to the PR owner.

Reviewer 4: True. I was just thinking that it might be worth recommending the apprach "edit a link into the PR description, which points to a comment with a summary in the PR conversation".

Author: Yeah I can add it.

Author: Clarified.

> * Consolidating the perspectives and putting a summary in the PR.

Reviewer 4: One more thing I'm wondering about: Should I make sure that my PR closes an issue? I.e. is it OK to just create a PR without a corresponding issue? If it's OK, we need to ensure that such PRs are accordingly triaged (required-for-ga, priority, ...).

Author: Personally I think it should be okay for people to create PR without a corresponding issue.

* Tagging subdomain experts (by looking at the change history) in the PR asking
  for suggestion.
* Reaching out to more people on the [gitter
  channel](https://gitter.im/open-telemetry/opentelemetry-specification).

> * Stepping back to see if it makes sense to scope down the PR.

Reviewer 4: ```suggestion
* Stepping back to see if it makes sense to narrow down the scope of the PR or split it up.
```

```bash
make misspell-correction
```

## Pull Requests

### How to create a PR

Everyone is welcome to contribute to the OpenTelemetry specification via GitHub
pull requests (PRs).

To [create a new
PR](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request),
fork the project in GitHub and clone the upstream repo:

```sh
git clone https://github.com/open-telemetry/opentelemetry-specification.git
```

Add your fork as a remote:

```sh
git remote add fork https://github.com/YOUR_GITHUB_USERNAME/opentelemetry-specification.git
```

Check out a new branch, make modifications and push the branch to your fork:

```sh
$ git checkout -b feature
# edit files
$ git commit
$ git push fork feature
```

Open a pull request against the main `opentelemetry-specification` repo.

If the PR is not ready for review, please mark it as
[`draft`](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

Please make sure CLA is signed and CI is clear. We don't expect people to review
and comment on a PR that doesn't have CLA signed.

### How to get your PR merged

A PR is considered to be **ready to merge** when:

> * It has received more than two approvals from the [code

Reviewer 2: IRC the actual requirement is that the author and approvers need to work in different companies, i.e. author works at X and approvers can both work in Y.

Author: I disagree with this, we have folks contributing PR but we don't really know which company they are working for.
We do have good understanding on what companies do the approvers work for.

Reviewer 2: I see your point. I'm personally fine with this approach.

Reviewer 2: (Worth changing it for all the SIGs, if we want to go this way)

>   owners](./.github/CODEOWNERS) (if approvals are from only one company, it

Reviewer 4: What is "it" here? The approvals?

```suggestion
  owners](./.github/CODEOWNERS) (if approvals are from only one company, they
```

>   owners](./.github/CODEOWNERS) (if approvals are from only one company, they

Reviewer 5: This is hard to enforce by GitHub, but I am fine with this in general

Reviewer 2: Agreed. We can look into this in the medium future.

  won't count).

> * There is no `request changes` from the [code owners](./.github/CODEOWNERS).

Reviewer 6: It should be possible to dismiss a change request if concerns were addressed or reasons not to do so were justified and there hasn't been any response to that by the reviewer for more than X days (where X should be somewhere between 7 and 14).

Reviewer 7: Does GitHub have such a functionality? Or can maintainer override request for change?

Reviewer 6: @iNikem Yes, both approving and blocking reviews can be dismissed by requesting a re-review (not sure what permissions are required for that) and repo admins (= maintainers/TC for spec) can merge anyway regardless of any reviews.

![image](https://user-images.githubusercontent.com/7052238/89032300-aa786b00-d334-11ea-88ba-d859a63c53f4.png)

Reviewer 4: > can be dismissed by requesting a re-review

No, that just re-requests a review. The dismissal is in the box at the bottom.

![image](https://user-images.githubusercontent.com/849039/89032425-eb707f80-d334-11ea-9587-4f40763b303a.png)

Reviewer 6: Ah, you're right. That also seems to work for blocking reviews:
![image](https://user-images.githubusercontent.com/7052238/89032531-296da380-d335-11ea-9c0b-3756f842c0c5.png)

Reviewer 2: Marking this issue as resolved to get this PR moving forward faster. @arminru Please open an issue if/when/as needed in order to do a follow up ;)

> * It has been at least two working days since the last modification (except for

Reviewer 2: I'd personally change to 1 day during this season, as we are trying to work hard towards GA. Once we reached stability, we can wait for days ;)

Author: I put two working days as I wish to put time zone / public holiday into consideration.
For example, a whole working day in UTC+1100 might have almost no overlap with UTC-1100.

Reviewer 2: Fair enough.

> * It has been at least two working days since the last modification (except for

Reviewer 6: I'd also add that any significant change should be clearly mentioned in a separate comment to not go unnoticed.
If the PR turns into a different direction after approving reviews, those should be dismissed and asked for a re-review.

Reviewer 2: (Same as the previous comment)

  the trivial updates, such like typo, cosmetic, rebase, etc.). This gives
  people reasonable time to review.
* Trivial changes (typos, cosmetic changes, CI improvements, etc.) don't have to
  wait for two days.

Any [spec

> maintainers](https://github.com/open-telemetry/community/blob/master/community-members.md#specifications-and-proto) can

Reviewer 4: ```suggestion
maintainer](https://github.com/open-telemetry/community/blob/master/community-members.md#specifications-and-proto) can
```

merge the PR once it is **ready to merge**.

If a PR has been stuck (e.g. there are lots of debates and people couldn't agree

> on each other), the owner should try to get people aligned by:

Reviewer 3: Can we add language here that maintainers should also be driving this, not only the PR owner? I think popular OSS projects tend to have maintainers that are proactive about making sure contributors have a good experience, and especially here where it's often maintainers debating points with each other, it's not really a good experience to leave all the management to the contributor.

Reviewer 2: +1 to have both maintainers and owners, specially if they are OTel organization members.

Reviewer 2: Let's do this in a follow up.

> on each other), the owner should try to get people aligned by:

Reviewer 4: ```suggestion
with each other), the owner should try to get people aligned by:
```
